<a href="https://colab.research.google.com/github/atoothman/data71200/blob/main/DATA_71200__Project_1__Wine.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DATA 71200 - Project 1 Wine

The Wine dataset used was found on the UC Irvine Machine Learning Repository

**Insert background and findings from this**

# Step 1: Load Dataset

Machine Learning Repository https://archive.ics.uci.edu/dataset/109/wine

In [20]:
import pandas as pd

# URL of the dataset
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data"


# Load the dataset into DataFrame
wine_df = pd.read_csv(url, header=None, index_col=False, names=[
                'Class', 'Alcohol', 'Malic acid', 'Ash', 'Alcalinity of ash',
                'Magnesium', 'Total phenols', 'Flavanoids', 'Nonflavanoid phenols',
                'Proanthocyanins', 'Color intensity', 'Hue', 'OD280/OD315 of diluted wines',
                'Proline'])

# Display wine_data DF and check that all features are listed
display(wine_df)

Class  Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0        1    14.23        1.71  2.43               15.6        127   
1        1    13.20        1.78  2.14               11.2        100   
2        1    13.16        2.36  2.67               18.6        101   
3        1    14.37        1.95  2.50               16.8        113   
4        1    13.24        2.59  2.87               21.0        118   
..     ...      ...         ...   ...                ...        ...   
173      3    13.71        5.65  2.45               20.5         95   
174      3    13.40        3.91  2.48               23.0        102   
175      3    13.27        4.28  2.26               20.0        120   
176      3    13.17        2.59  2.37               20.0        120   
177      3    14.13        4.10  2.74               24.5         96   

     Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0             2.80        3.06                  0.28             2.29   
1             2.65        2.76                  0.26             1.28   
2             2.80        3.24                  0.30             2.81   
3             3.85        3.49                  0.24             2.18   
4             2.80        2.69                  0.39             1.82   
..             ...         ...                   ...              ...   
173           1.68        0.61                  0.52             1.06   
174           1.80        0.75                  0.43             1.41   
175           1.59        0.69                  0.43             1.35   
176           1.65        0.68                  0.53             1.46   
177           2.05        0.76                  0.56             1.35   

     Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0               5.64  1.04                          3.92     1065  
1               4.38  1.05                          3.40     1050  
2               5.68  1.03                          3.17     1185  
3               7.80  0.86                          3.45     1480  
4               4.32  1.04                          2.93      735  
..               ...   ...                           ...      ...  
173             7.70  0.64                          1.74      740  
174             7.30  0.70                          1.56      750  
175            10.20  0.59                          1.56      835  
176             9.30  0.60                          1.62      840  
177             9.20  0.61                          1.60      560  

[178 rows x 14 columns]

Confirming that all features (13) and instances (178) as listed on the website have been loaded.

# Explore the dataset
*note that class is a categorical variable, but has Dtype as int below

In [46]:
# Show info on wine_df
print("Info on wine_df:")
wine_df.info()
print()


# Show description of wine_df
print("Description of wine_df:")
wine_df.describe()

Info on wine_df:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Class                         178 non-null    int64  
 1   Alcohol                       178 non-null    float64
 2   Malic acid                    178 non-null    float64
 3   Ash                           178 non-null    float64
 4   Alcalinity of ash             178 non-null    float64
 5   Magnesium                     178 non-null    int64  
 6   Total phenols                 178 non-null    float64
 7   Flavanoids                    178 non-null    float64
 8   Nonflavanoid phenols          178 non-null    float64
 9   Proanthocyanins               178 non-null    float64
 10  Color intensity               178 non-null    float64
 11  Hue                           178 non-null    float64
 12  OD280/OD315 of diluted wines  178 non-null    f

Class     Alcohol  Malic acid         Ash  Alcalinity of ash  \
count  178.000000  178.000000  178.000000  178.000000         178.000000   
mean     1.938202   13.000618    2.336348    2.366517          19.494944   
std      0.775035    0.811827    1.117146    0.274344           3.339564   
min      1.000000   11.030000    0.740000    1.360000          10.600000   
25%      1.000000   12.362500    1.602500    2.210000          17.200000   
50%      2.000000   13.050000    1.865000    2.360000          19.500000   
75%      3.000000   13.677500    3.082500    2.557500          21.500000   
max      3.000000   14.830000    5.800000    3.230000          30.000000   

        Magnesium  Total phenols  Flavanoids  Nonflavanoid phenols  \
count  178.000000     178.000000  178.000000            178.000000   
mean    99.741573       2.295112    2.029270              0.361854   
std     14.282484       0.625851    0.998859              0.124453   
min     70.000000       0.980000    0.340000              0.130000   
25%     88.000000       1.742500    1.205000              0.270000   
50%     98.000000       2.355000    2.135000              0.340000   
75%    107.000000       2.800000    2.875000              0.437500   
max    162.000000       3.880000    5.080000              0.660000   

       Proanthocyanins  Color intensity         Hue  \
count       178.000000       178.000000  178.000000   
mean          1.590899         5.058090    0.957449   
std           0.572359         2.318286    0.228572   
min           0.410000         1.280000    0.480000   
25%           1.250000         3.220000    0.782500   
50%           1.555000         4.690000    0.965000   
75%           1.950000         6.200000    1.120000   
max           3.580000        13.000000    1.710000   

       OD280/OD315 of diluted wines      Proline  
count                    178.000000   178.000000  
mean                       2.611685   746.893258  
std                        0.709990   314.907474  
min                        1.270000   278.000000  
25%                        1.937500   500.500000  
50%                        2.780000   673.500000  
75%                        3.170000   985.000000  
max                        4.000000  1680.000000

# Checking for missing values in dataset before splitting

The dataset on UCI ML repository list zero missing values. Double checking below and confirming there are zero missing values.

In [22]:
# Check for missing values in wine_df
missing_values = wine_df.isnull().sum()

# Print missing values in wine_df
print("Missing Values:")
print(missing_values)

Missing Values:
Class                           0
Alcohol                         0
Malic acid                      0
Ash                             0
Alcalinity of ash               0
Magnesium                       0
Total phenols                   0
Flavanoids                      0
Nonflavanoid phenols            0
Proanthocyanins                 0
Color intensity                 0
Hue                             0
OD280/OD315 of diluted wines    0
Proline                         0
dtype: int64


# Step 2: Divide into a training set and a testing set

To create a balanced split StratifiedShuffleSplit will be utilized below

In [32]:
# Utilize scikitlearn for a stratify split
from sklearn.model_selection import StratifiedShuffleSplit

# Separate the features (X) and the target variable (y) aka Class
X = wine_df.drop('Class', axis=1)
y = wine_df['Class']

# Create StratifiedShuffleSplit with 1 split, 20% test size, 80% training and random state 42
split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
# Split data
for train_index, test_index in split.split(X, y):
   X_train = X.iloc[train_index]
   X_test = X.iloc[test_index]
   y_train = y.iloc[train_index]
   y_test = y.iloc[test_index]

# Check to confirm successful 80 : 20 split
print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (142, 13)
X_test: (36, 13)
y_train: (142,)
y_test: (36,)


Confirming the split was successful. The training set has 80% of the original 178 instances, while the testing set has 20% of the original instances. 13 in the X-train and X_test show the number of columns/features.

# Step 3: Explore your training set

with Pandas data fram and using the below panadas funcations to explore data
*   DataFrame.info()
*   DataFrame.describe()

In [44]:
# Training set infomation on X_train
print("Info on X_train:")
X_train.info()
print()


# Training set description of X_train
print("Description of X_train:")
X_train.describe()

Info on X_train:
<class 'pandas.core.frame.DataFrame'>
Index: 142 entries, 36 to 131
Data columns (total 13 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Alcohol                       142 non-null    float64
 1   Malic acid                    142 non-null    float64
 2   Ash                           142 non-null    float64
 3   Alcalinity of ash             142 non-null    float64
 4   Magnesium                     142 non-null    int64  
 5   Total phenols                 142 non-null    float64
 6   Flavanoids                    142 non-null    float64
 7   Nonflavanoid phenols          142 non-null    float64
 8   Proanthocyanins               142 non-null    float64
 9   Color intensity               142 non-null    float64
 10  Hue                           142 non-null    float64
 11  OD280/OD315 of diluted wines  142 non-null    float64
 12  Proline                       142 non-null    int64

Alcohol  Malic acid         Ash  Alcalinity of ash   Magnesium  \
count  142.000000  142.000000  142.000000         142.000000  142.000000   
mean    12.971479    2.340000    2.364859          19.625352   99.633803   
std      0.802521    1.101285    0.268380           3.380848   14.937448   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.370000    1.610000    2.200000          17.200000   88.000000   
50%     13.010000    1.810000    2.360000          19.500000   97.000000   
75%     13.560000    3.170000    2.560000          21.500000  107.750000   
max     14.830000    5.800000    3.220000          30.000000  162.000000   

       Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
count     142.000000  142.000000            142.000000       142.000000   
mean        2.273732    1.985352              0.360141         1.600211   
std         0.621454    0.951952              0.119211         0.579777   
min         0.980000    0.340000              0.130000         0.420000   
25%         1.705000    1.227500              0.270000         1.250000   
50%         2.335000    2.135000              0.340000         1.560000   
75%         2.800000    2.775000              0.430000         1.950000   
max         3.880000    3.740000              0.630000         3.580000   

       Color intensity         Hue  OD280/OD315 of diluted wines      Proline  
count       142.000000  142.000000                    142.000000   142.000000  
mean          4.989648    0.948986                      2.606197   739.478873  
std           2.334947    0.226647                      0.689650   301.499014  
min           1.280000    0.480000                      1.270000   278.000000  
25%           3.057500    0.780000                      1.970000   496.250000  
50%           4.640000    0.960000                      2.780000   673.500000  
75%           6.182500    1.090000                      3.160000   981.250000  
max          13.000000    1.710000                      3.920000  1515.000000

Class     Alcohol  Malic acid         Ash  Alcalinity of ash  \
count  178.000000  178.000000  178.000000  178.000000         178.000000   
mean     1.938202   13.000618    2.336348    2.366517          19.494944   
std      0.775035    0.811827    1.117146    0.274344           3.339564   
min      1.000000   11.030000    0.740000    1.360000          10.600000   
25%      1.000000   12.362500    1.602500    2.210000          17.200000   
50%      2.000000   13.050000    1.865000    2.360000          19.500000   
75%      3.000000   13.677500    3.082500    2.557500          21.500000   
max      3.000000   14.830000    5.800000    3.230000          30.000000   

        Magnesium  Total phenols  Flavanoids  Nonflavanoid phenols  \
count  178.000000     178.000000  178.000000            178.000000   
mean    99.741573       2.295112    2.029270              0.361854   
std     14.282484       0.625851    0.998859              0.124453   
min     70.000000       0.980000    0.340000              0.130000   
25%     88.000000       1.742500    1.205000              0.270000   
50%     98.000000       2.355000    2.135000              0.340000   
75%    107.000000       2.800000    2.875000              0.437500   
max    162.000000       3.880000    5.080000              0.660000   

       Proanthocyanins  Color intensity         Hue  \
count       178.000000       178.000000  178.000000   
mean          1.590899         5.058090    0.957449   
std           0.572359         2.318286    0.228572   
min           0.410000         1.280000    0.480000   
25%           1.250000         3.220000    0.782500   
50%           1.555000         4.690000    0.965000   
75%           1.950000         6.200000    1.120000   
max           3.580000        13.000000    1.710000   

       OD280/OD315 of diluted wines      Proline  
count                    178.000000   178.000000  
mean                       2.611685   746.893258  
std                        0.709990   314.907474  
min                        1.270000   278.000000  
25%                        1.937500   500.500000  
50%                        2.780000   673.500000  
75%                        3.170000   985.000000  
max                        4.000000  1680.000000

In [45]:
# Training set infomation on y_train
print("Info on y_train:")
y_train.info()
print()

# Training set description of y_train
print("Description of y_train:")
y_train.describe()

Info on y_train:
<class 'pandas.core.series.Series'>
Index: 142 entries, 36 to 131
Series name: Class
Non-Null Count  Dtype
--------------  -----
142 non-null    int64
dtypes: int64(1)
memory usage: 2.2 KB

Description of y_train:


count    142.000000
mean       1.936620
std        0.773816
min        1.000000
25%        1.000000
50%        2.000000
75%        3.000000
max        3.000000
Name: Class, dtype: float64

# Step 4: Data Cleaning

Address any missing values in the training set and testing set

[22] confirmed there was no missing values in the wine_df. The below is to ensure there are no missing values in the training or testing sets.

In [53]:
# Checking missing values in training set

# Check for missing values in X_train
missing_values = X_train.isnull().sum()

# Print missing values in X_train
print("Missing Values in X_train:")
print(missing_values)
print()

# Check for missing values in y_train
missing_values = y_train.isnull().sum()

# Print missing values in y_train
print("Missing Values in y_train:")
print(missing_values)

Missing Values in X_train:
Alcohol                         0
Malic acid                      0
Ash                             0
Alcalinity of ash               0
Magnesium                       0
Total phenols                   0
Flavanoids                      0
Nonflavanoid phenols            0
Proanthocyanins                 0
Color intensity                 0
Hue                             0
OD280/OD315 of diluted wines    0
Proline                         0
dtype: int64

Missing Values in y_train:
0


In [54]:
# Checking missing values in test set

# Check for missing values in X_test
missing_values = X_test.isnull().sum()

# Print missing values in X_test
print("Missing Values in X_test:")
print(missing_values)
print()

# Check for missing values in y_test
missing_values = y_test.isnull().sum()

# Print missing values in y_test
print("Missing Values in y_test:")
print(missing_values)

Missing Values in X_test:
Alcohol                         0
Malic acid                      0
Ash                             0
Alcalinity of ash               0
Magnesium                       0
Total phenols                   0
Flavanoids                      0
Nonflavanoid phenols            0
Proanthocyanins                 0
Color intensity                 0
Hue                             0
OD280/OD315 of diluted wines    0
Proline                         0
dtype: int64

Missing Values in y_test:
0


# Step 5: Visualize the data in the training set

Utilize the pandas functions to visualize the data

*   DataFrame.hist
*   plotting.scatter_matrix()

X_train

y_train


In [57]:
%matplotlib inline
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix

In [61]:
# Histogram of original wine_df
# wine_df.hist(bins=50, figsize=(20,15))
# plt.show()

# Scatter_matrix of original wine_df
# attributes = ['Class', 'Alcohol', 'Malic acid', 'Ash', 'Alcalinity of ash',
               # 'Magnesium', 'Total phenols', 'Flavanoids', 'Nonflavanoid phenols',
               # 'Proanthocyanins', 'Color intensity', 'Hue', 'OD280/OD315 of diluted wines',
               # 'Proline']
# scatter_matrix(wine_df[attributes], figsize=(12, 8))

# Step 6: Apply transformations to data

Apply the below transofrmation to 2 features from the dataset
*  Squaring (**2)
*  Cubing (**3)
*  np.log
*  np.exp

Plot histograms and scatter matrices to explore the resultant data